prv_accountant/composers.py

```diff
@@ -67,3 +70,75 @@ def compute_composition(self, num_compositions: Sequence[int]) -> DiscretePrivac
         domain = self.domain.shift_right(self.domain.shifts()*(num_compositions-1))
 
         return DiscretePrivacyRandomVariable(f_n, domain)
+
+
+class ConvolutionTree(Composer):
+    def __init__(self, prvs: Sequence[DiscretePrivacyRandomVariable]) -> None:
+        """
+        Create a composer for efficiently composing heterogeneous PRVs.
+
+        This composer runs in $n \\log n$ where $n$ is the total number of compositions. Hence, it isn't optimal for
+        homogeneous composition. Use the Fourier composer for homogeneous composition instead.
+
+        :param Sequence[DiscretePrivacyRandomVariable] prvs: Sequence of discrete PRVs to compose
+        """
+        super().__init__(prvs=prvs)
+
+    def compute_composition(self, num_self_compositions: Sequence[int]) -> DiscretePrivacyRandomVariable:
+        """
+        Compute the composition of PRVs.
+
+        Since this composer is not efficient for homogeneous composition. `num_self_composition` is expected to be 1.
+
+        :param num_self_compositions: The number of composition for each PRV with itself.
+
+        :type num_self_compositions: Sequence[int]
+        """
+        if (np.array(num_self_compositions)-1).any():
+            raise ValueError("Cannot handle homogeneous composition. Use Fourier composer for that.")
+        prvs = self.prvs
+        while len(prvs) > 1:
+            if len(prvs) % 2 == 1:
+                prvs_conv = [prvs.pop(0)]
+            else:
+                prvs_conv = []
+            for prv_L, prv_R in zip(prvs[:-1:2], prvs[1::2]):
+                prvs_conv.append(self._add_prvs(prv_L, prv_R))
+            prvs = prvs_conv
+        return prvs[0]
+
+    def _add_prvs(self, prv_L: DiscretePrivacyRandomVariable,
+                  prv_R: DiscretePrivacyRandomVariable) -> DiscretePrivacyRandomVariable:
+        f = convolve(prv_L.pmf, prv_R.pmf, mode="same")
+        domain = prv_L.domain.shift_right(prv_R.domain.shifts())
+        return DiscretePrivacyRandomVariable(f, domain)
+
+
+class Heterogeneous(Composer):
```

Comment:
I think that could make sense. I think it would fit in the current structure quite nicely. But I'm a bit hesitant that we might be too restrictive if we want to create some hybrid composer at some point.